File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
str = input('')

while(str != "."):
    stack = []
    bal = True
    for a in str:
        if(a == '[' or a == '('):
            stack.append(a)
        elif(a == ']'):
            if( (len(stack) == 0) or stack[-1] != '[' ):
                bal = False
            else:
                stack.pop()
        elif(a == ')'):
            if( (len(stack) == 0) or stack[-1] != '(' ):
                bal = False
            else:
                stack.pop()

        if(not(bal)):
            break
    
    if(not(bal) or (len(stack) != 0)):
        print('no')
    else:
        print('yes')

    str = input()

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.cc
#include <iostream>
#include <stack>
#include <string>
using namespace std;

int main()
{
	stack<char> s;
	string str;

	do
	{
	NEXT:
		while (!s.empty())
			s.pop();
		getline(cin, str, '.');
		cin.ignore();
		if (str == "")
			return 0;
			

		while ( !str.empty() )
		{
			char a = str.front();
			switch (a)
			{
			case '[':
			case '(':
				s.push(a);
				break;


			case ']':
				if ( s.empty() || (s.top() != '[') )
				{
					cout << "no\n";
					goto NEXT;
				}
				s.pop();
				break;

			case ')':
				if ( s.empty() || (s.top() != '(') )
				{
					cout << "no\n";
					goto NEXT;
				}
				s.pop();
				break;


			default:
				break;
			}
			str.erase(str.begin());
		}

		if(s.empty())
			cout << "yes\n";
		else
			cout << "no\n";

	} while (true);

	return 0;
}